Feature: Enable appending to trajectory when using `ts.optimize`/`ts.integrate`

danielzuegner opened the pull request:
## Summary

* Adds functionality to read the latest step per trajectory and trajectory reporter to set the initial step accordingly.
* Adds tests `test_optimize_append_to_trajectory` and `test_integrate_append_to_trajectory` to ensure trajectories are being appended to.
* Also fixes a bug where, whenever a structure has converged, all other trajectories are reset by re-opening them in `"w"` mode: https://github.com/TorchSim/torch-sim/blob/main/torch_sim/runners.py#L483.

Closes https://github.com/TorchSim/torch-sim/issues/356.
## Checklist

Before a pull request can be merged, the following items must be checked:

* [x] Doc strings have been added in the [Google docstring format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html#example-google).
* [x] Run [ruff](https://beta.ruff.rs/docs/rules/#pydocstyle-d) on your code.
* [x] Tests have been added for any new functionality or bug fixes.

<!-- We highly recommended installing the `prek` hooks running in CI locally to speedup the development process. Simply run `pip install prek && prek install` to install the hooks which will check your code before each commit. -->

orionarcher commented on torch_sim/runners.py:
```diff
@@ -485,9 +501,13 @@ def optimize[T: OptimState](  # noqa: C901, PLR0915
         ):
+            mode_before = trajectory_reporter.trajectory_kwargs["mode"]
+            # temporarily set to "append" mode to avoid overwriting existing files
+            trajectory_reporter.trajectory_kwargs["mode"] = "a"
```
Is the append mode fixing the bug or just necessary for appending?

More generally, this isn't grappling with the fact that some the existing trajectories might already exist or might have different numbers of steps

thomasloux commented on tests/test_trajectory.py:
```diff
@@ -834,3 +835,99 @@ def test_write_ase_trajectory_importerror(
+        _ = ts.integrate(
+            system=int_state,
+            model=lj_model,
+            timestep=0.1,
```
This timestep is really big as the unit is ps, consider using 0.01 or 0.001

danielzuegner commented on torch_sim/trajectory.py:
```diff
@@ -302,6 +301,43 @@ def close(self) -> None:
+            else:
+                with TorchSimTrajectory(trajectory._file.filename, mode="r") as traj:
+                    last_step = traj.last_step
+            max_step = max(max_step, last_step)
```
Please take a look at the current implementation, it turned out to be not so difficult to implement. Now there's no "gap" anymore in the trajectories.

thomasloux commented:
For a MD simulations at least, there is no such thing, all systems evolve at the same time. I actually assumes it the same for optimize, if not for convergence criteria that would be met.
So in case of `ts.integrate`,  you would get `n_step` from trajectory, and you would run a for loop like `range(n_step_traj, n_step_total)`instead of the current PR suggestion `range(n_initial, n_initial + n_step)` and then having to manually remove n_initial when you want to access kT values for instance.

thomasloux commented:
I don't have any strict opinion on optimization as I don't use it this much

thomasloux commented on torch_sim/runners.py:
```diff
@@ -499,14 +538,21 @@ def optimize[T: OptimState](  # noqa: C901, PLR0915
+                    state, (initial_step + steps_so_far).tolist(), model=model
+                )
+            steps_so_far += 1
+            exceeded_max_steps = (initial_step + steps_so_far) > max_steps
```
I did not run this code, but I highly suspect this is wrong. you add steps_so_far, that is a global counter of the while loop, while initial_step is a slice of the og_initial_step. This means, if state is large enough to be divided in multiple batches, you still account for the step of batch 1...n-1 for batch n. You should to `step = og_initial_step[autobatcher.current_idx]`and then incremement `step += 1`. This should increment the og_initial_step so that you are consistent on the step of each one.

thomasloux reviewed (changes requested):
One part to clarify for `ts.optimize`, otherwise it's a really nice work. I think in a later PR we should clarify the Trajectory objects. I'm not a fan of the mode 'r' vs 'a'. I would probably prefer an argument: either 'restart=True' or 'overwrite_if_exists=True'.

thomasloux reviewed (approved):
good for me now

orionarcher commented on torch_sim/trajectory.py:
```diff
@@ -141,19 +140,39 @@ def __init__(
+        """
+        if not self.trajectories:
+            return None
+        return [traj.filename for traj in self.trajectories]
```
This is a good change

orionarcher reviewed (changes requested):
>  I've updated the code to assume that for integrate() all trajectories start from the same time step, and then only integrate for n_steps - initial_step steps. Is that in line with what you had in mind?

Sorry for being a bit late on this, but I am going to push back on this implementation.

I think that n_steps should mean the number of steps you take, not the number of steps you reach. My heavy prior would be that n_steps is a number of steps taken, as a user, I'd find it quite strange if that wasn't the case. What do you think?

I also don't think that if the steps don't match that we should truncate all the trajectories to the same length. In my mind, we should detect the current step across all the trajectories, then progressively append to that step + i, where i is the number of steps taken by the integrator. Does that make sense?

I want to say I am quite appreciative of the PR and I think being able to append with the runners is a good feature. I am being nit picky because I think it's challenging to get the right behavior and expectations for this in a batched context.

thomasloux commented:
Hey Orion, I think that both implementations make sense. I see two possibilities:
- The user has a mode="a" in mind, so it makes sense to have n_steps to be the n_steps to add and not the total number of steps in trajectory.
- The user wants to restart a trajectory (think of a failure or a slurm job that finished too early). Then I think it makes sense to run the same code and the that TorchSim can detect the traj file to start from it.

I think the 2nd option is more appealing then the first one. Also we could easily obtained the first behaviour by adding an argument to the functions.

The truncation is also a direct result of the fact I pushed for a restart behaviour. At the moment a MD simulation from ts.integrate should always be in sync for all states in batch, so it does not make sense to allow different indices in this PR. We could change that in another PR though

orionarcher commented:
Good framing. Both use cases are valid so perhaps it's worth considering the left out group. 

### Finishing the simulation if `n_steps` = additional steps

```python
total_steps = 1000
steps_completed_before_interrupt = my_trajectory_reporter.last_step  # using the last_step property introduced in this PR
n_steps = 1000 - steps_completed_before_interupt
integrate(...n_steps=n_steps)
```

### Appending 20 steps if `n_steps` = total steps

```python
new_steps = 1000
current_number_of_steps = my_trajectory_reporter.last_step
n_steps = current_number_of_steps - new_steps
integrate(...n_steps=n_steps)
```

Either is still accessible no matter what default is adopted and both are about the same amount of code. Ultimately, it comes down to which is more expected and which is more common. I am not sure which is more common, it very well might be restarting, but I feel that appending steps is more expected of an `n_steps` argument. 

> At the moment a MD simulation from ts.integrate should always be in sync for all states in batch, so it does not make sense to allow different indices in this PR. We could change that in another PR though

Good point that separate indices are a distinct feature and could go elsewhere. This is a potentially dangerous default though. Imagine I have 100 simulations and I can fit 10 on a GPU at a time. My simulation proceeds nicely through the first 90 and then only makes it 1 step into the the final 10. If I rerun the same script expecting to restart, I'll erase the progress on the first 90 simulations because the smallest `last_step` will be 1. Instead of truncating, perhaps we could enforce that all trajectories have the same `last_step`? That feels safer to me.

> The user wants to restart a trajectory (think of a failure or a slurm job that finished too early). Then I think it makes sense to run the same code and the that TorchSim can detect the traj file to start from it.

This is going to be difficult when some simulations will run to completion and others will not begin. I think some modifications will need to be made to the script to weed out already started PRs. 


OK, as I outline these challenges and think about it I am realizing that it's pretty hard to restart simulations in TorchSim right now and my proposal would keep it hard. Perhaps we can add some flag like `restart=False` in the kwargs that will nicely handle the complexities outlined above and allow a script to be rerun without worrying about any of this. That would require a bit of logic added to what we have now. I have to run right now but will think on this further.

danielzuegner commented on torch_sim/trajectory.py:
```diff
@@ -167,19 +186,23 @@ def load_new_trajectories(
+            for filename in filenames
+        ]
+        # Restore original mode
+        self.trajectory_kwargs["mode"] = _mode
```
My argument is that the mode (`"w"` vs `"a"`) is meant by the user only for the initial opening of the file, not the (potentially many) times we re-open the trajectory file during the optimization (e.g., when some of the trajectories finish). I don't think it is ever intentional to wipe all existing progress when that happens, even if the user specified `"w"` mode in the beginning.

Hence, in my PR I propose to use the user-specified mode only for the first time we open the trajectory files (in the constructor of the trajectory reporter), and afterwards avoid wiping then files by ensuring `"a"` mode for the times we re-open the files during the optimization.

What is your alternative proposal to ensure that (i) we use the user-intended file mode initially, but (ii) don't erase the files during the optimization?

orionarcher replied:
Setting aside where it's intentional, I do think it should be possible to wipe the existing files when re-running, though I agree that by default appending is better. LAMMPS has an append mode when writing dump files that is by [default set to `False`](https://docs.lammps.org/dump.html), so this particular sharp edge isn't unusual.

I would propose that we set the default to "a" but wipe the files if "w" is set.

danielzuegner commented on torch_sim/trajectory.py:
```diff
@@ -255,18 +280,19 @@ def report(
         all_props: list[dict[str, torch.Tensor]] = []
         # Process each system separately
         for idx, substate in enumerate(split_states):
+            _step = step[idx] if isinstance(step, list) else step
```
thanks, done!

danielzuegner commented:
@orionarcher could you take another look? I've changed the behavior to always append `n_steps` in the case of `integrate`. For `optimize` I've kept the behavior as-is because the argument is `max_steps` anyway. I've also added `TrajectoryReporter.truncate_to_step` to provide a higher-level API to make trajectories consistent in terms of number of steps.

The test failures here appear to be unrelated to this PR, feel free to let me know if you think otherwise.

Thanks!

danielzuegner commented:
@orionarcher ping :)

orionarcher reviewed (approved):
Nicely done @danielzuegner! Thank you for your persistence in completing this PR. This is a very nice addition to the high-level runners.

Just a couple small things but I trust you can resolve them.
